### PR TITLE
Feature/column config dataframe guide

### PR DESCRIPTION
**Pull request by sfc-gh-dmatthews**
## 📚 Context
This commit completes the documentation for the new feature column config.

## 🧠 Description of Changes
This commit updates the Dataframe guide in Advanced features to include an overview of column configuration. It refines the API pages on `st.dataframe` and `st.data_editor` to reduce duplicate content. An app to visually demonstrate column configuration along with a brief summary redirects users back to the Column configuration API documentation.

Additionally, this commit removes a KB article describing a workaround to hide dataframe indices. This is now a built-in feature. A new KB article describes a workaround to get user row-selections until that becomes a built-in feature as well.

**Revised:**
Will link to preview.

## 💥 Impact
Size:
- [X] Not small

## 🌐 References
- Notion: [Sprint task](https://www.notion.so/snowflake-corp/Feature-docs-for-6f09ae87039e4620ab3a504812fa7beb)
- Notion: [Spec 1](https://www.notion.so/snowflake-corp/Doc-spec-for-advanced-features-dataframes-re-column-config-b88901fd01f5472d9178893ae39a58eb?pvs=4)
- Notion: [Spec 2](https://www.notion.so/snowflake-corp/New-section-ideas-on-dataframes-b036bdd716e64be09d5e06f6b907326f?pvs=4)

**Contribution License Agreement**
By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

### CODE REVIEW

File: content/library/advanced-features/dataframes.md
Comment (sfc-gh-dmatthews) [inline content/library/advanced-features/dataframes.md:287]: The current requirements file has an older version of pandas pinned. I wrote this app to use pandas>=2.0 since that's what installs by default for all versions of streamlit with column configuration. Can we unpin the version of pandas so it installs the latest? Or pin it to a higher version? Do we have any embedded examples relying on something deprecated in pandas?

Comment (sfc-gh-dmatthews) [reply content/library/advanced-features/dataframes.md:287]: I have updated the pandas version to above 2.0.

File: content/kb/using-streamlit/how-to-get-row-selections.md
Comment (jrieke) [inline content/kb/using-streamlit/how-to-get-row-selections.md:33]: I would leave out `:pd:DataFrame` here, I think that's quite confusing for beginners. And doesn't seem to add too much value since it should be pretty clear from the name `edited_df` that this is a dataframe.

File: content/library/advanced-features/dataframes.md
Comment (jrieke) [inline content/library/advanced-features/dataframes.md:235]: Might be a bit easier to read if it's:

    Column configuration includes the following column types: Text, Number, ...

File: content/library/advanced-features/dataframes.md
Comment (jrieke) [inline content/library/advanced-features/dataframes.md:241]: I don't think we should push people towards pandas Styler here, since we want to get people away from it as much as possible. We might even get rid of it completely at some point in the future. I would not mention it here.

File: content/library/advanced-features/dataframes.md
Comment (jrieke) [inline content/library/advanced-features/dataframes.md:243]: Might make sense to link to the API docs here.

File: content/library/advanced-features/dataframes.md
Comment (jrieke) [inline content/library/advanced-features/dataframes.md:249]: I'm OK with listing them out explicitly as here, but I could also imagine making it a bit shorter and just saying something like "e.g. for text columns, you can set a maximum number of characters through `max_chars` or use regex validation in `validate`". Feels a bit like we are just repeating the API docs here...

File: content/library/advanced-features/dataframes.md
Comment (jrieke) [inline content/library/advanced-features/dataframes.md:291]: That seems super edge casey, not sure if we need to keep this in here. Maybe we could also move it to KB? Or document it on the API reference for st.data_editor? Or just leave it out altogether, given that more and more people will switch to pandas 2 eventually, anyway?

Comment (jrieke) [reply content/library/advanced-features/dataframes.md:291]: I think if this is an issue people have, they will just land on a forum post for this anyway...

**Review by jrieke — APPROVED** [review]
Only gave this a brief review. Left a few comments but overall LGTM. No need to get another review from me before merging. I may go through it again once it's live on the docs site but that doesn't have to block merging.

**Comment by jrieke** [comment]
PS: I LOOOVE the KB article on row selection!

File: content/kb/using-streamlit/how-to-get-row-selections.md
Comment (sfc-gh-dmatthews) [reply content/kb/using-streamlit/how-to-get-row-selections.md:33]: That was my bad. I think I added that when I was testing something and needed the type recognized. Sorry about that. Definitely will remove it.

File: content/library/advanced-features/dataframes.md
Comment (sfc-gh-dmatthews) [reply content/library/advanced-features/dataframes.md:235]: Agree.

File: content/library/advanced-features/dataframes.md
Comment (sfc-gh-dmatthews) [reply content/library/advanced-features/dataframes.md:241]: Since you've added the restriction to the docstring recently, I agree we can get rid of it here. I wanted to make sure that limitation was clear.

File: content/library/advanced-features/dataframes.md
Comment (sfc-gh-dmatthews) [reply content/library/advanced-features/dataframes.md:243]: Done.

File: content/library/advanced-features/dataframes.md
Comment (sfc-gh-dmatthews) [reply content/library/advanced-features/dataframes.md:249]: Concise for the win!

File: content/library/advanced-features/dataframes.md
Comment (sfc-gh-dmatthews) [reply content/library/advanced-features/dataframes.md:291]: I'll remove it, but also @LukasMasuch since he raised the issue originally.

**Review by sfc-gh-jrieke — APPROVED** [review]

**Pull request merged**
